examples/payload_scan/payload_scan.c

```diff
@@ -0,0 +1,266 @@
+
+        /* Clear screen and move to top left */
+        printf("%s%s", clr, topLeft);
+        printf("Search term: %s\n", search_term);
+        printf("Packets accepted: %lu\n", stats.pkt_accept);
+        printf("Packets dropped: %lu\n", stats.pkt_drop);
+        printf("Packets not IPv4: %lu\n", stats.pkt_not_ipv4);
+        printf("Packets not UDP/TCP: %lu\n", stats.pkt_not_tcp_udp);
+        printf("Packets total: %lu", stats.pkt_total);
+
+        printf("\n\n");
+}
+
+static int
+packet_handler(struct rte_mbuf *pkt, struct onvm_pkt_meta *meta,
+               __attribute__((unused)) struct onvm_nf_info *nf_info) {
+        int udp_pkt, tcp_pkt;
+        static uint32_t counter = 0;
+        uint8_t *pkt_data;
+
+        if (++counter == print_delay) {
+                do_stats_display();
+                counter = 0;
+        }
+
+        stats.pkt_total++;
+
+        if (!onvm_pkt_is_ipv4(pkt)) {
+                meta->action = ONVM_NF_ACTION_DROP;
+                stats.pkt_not_ipv4++;
+                return 0;
+        }
+
+        udp_pkt = onvm_pkt_is_udp(pkt);
+        tcp_pkt = onvm_pkt_is_tcp(pkt);
+        pkt_data = NULL;
+
+        if (!udp_pkt && !tcp_pkt) {
+                meta->action = ONVM_NF_ACTION_DROP;
+                stats.pkt_not_tcp_udp++;
+                return 0;
+        }
+
+        if (udp_pkt) {
+                pkt_data = rte_pktmbuf_mtod_offset(pkt, uint8_t * , sizeof(struct ether_hdr) + sizeof(struct ipv4_hdr) +
+                                                                    sizeof(struct udp_hdr));
+        } else {
+                pkt_data = rte_pktmbuf_mtod_offset(pkt, uint8_t * , sizeof(struct ether_hdr) + sizeof(struct ipv4_hdr) +
+                                                                    sizeof(struct tcp_hdr));
+        }
+
+        if (strstr((const char *) pkt_data, search_term) != NULL) {
+                if (!inverse) {
+                        meta->action = ONVM_NF_ACTION_TONF;
+                        meta->destination = destination;
+                        stats.pkt_accept++;
+
+                } else {
+                        meta->action = ONVM_NF_ACTION_DROP;
+                        stats.pkt_drop++;
+                }
+        } else {
+                if (!inverse) {
+                        meta->action = ONVM_NF_ACTION_DROP;
+                        stats.pkt_drop++;
+                } else {
+                        meta->action = ONVM_NF_ACTION_TONF;
+                        meta->destination = destination;
+                        stats.pkt_accept++;
+                }
+        }
+
+        return 0;
+}
+
+
+int main(int argc, char *argv[]) {
+        int arg_offset;
+        const char *progname = argv[0];
+
+        if ((arg_offset = onvm_nflib_init(argc, argv, NF_TAG, &nf_info)) < 0)
+                return -1;
+
+        argc -= arg_offset;
+        argv += arg_offset;
+
+        stats.pkt_drop = 0;
```

Comment:
Okay so regarding globals.
Globals are usually bad for NFs because if we ever want to scale we will end up sharing the global state which leads to race conditions between NFs. 

This is techincally not enforced in other NFs right now but the idea is putting the NF state data into the onvm_nf_info->data field, you will need to malloc your own memory instead of using the stack. Then the stats will be considered state data and are not going to be shared if we scale.

Reply:
Ok, the struct stats are now stored in the memory region nf_info->data. Let me know if that looks right..

Reply:
almost there